src/writer/_dirent.h

```diff
@@ -21,91 +21,150 @@
+      public: // structures
+        struct Direct {
+          Direct() :
+            cluster(nullptr),
+            blobNumber(0)
+          {};
+          Cluster*         cluster;
+          blob_index_t     blobNumber;
+        } PACKED;
+
+        struct Redirect {
+          Redirect(char ns, const std::string& target) :
+            targetPath(target),
+            ns(ns)
+          {};
+          Redirect(Redirect&& r) = default;
+          ~Redirect() {};
+          TinyString targetPath;
+          char ns;
+        } PACKED;
+
+        struct Resolved {
+          Resolved(const Dirent* target) :
+            targetDirent(target)
+          {};
+          const Dirent* targetDirent;
+        } PACKED;
+
+      public: // functions
+
+        ~DirentInfo() {
+          switch(tag) {
+            case DIRECT:
+             direct.~Direct();
+              break;
+            case REDIRECT:
+              redirect.~Redirect();
+              break;
+            case RESOLVED:
+              resolved.~Resolved();
+              break;
+          }
+        };
+        DirentInfo(Direct&& d):
+          direct(std::move(d)),
+          tag(DirentInfo::DIRECT)
+        {}
+        DirentInfo(Redirect&& r):
+          redirect(std::move(r)),
+          tag(DirentInfo::REDIRECT)
+        {}
+        DirentInfo(Resolved&& r):
+          resolved(std::move(r)),
+          tag(DirentInfo::RESOLVED)
+        {}
+        DirentInfo::Direct& getDirect() {
+          ASSERT(tag, ==, DIRECT);
+          return direct;
+        }
+        DirentInfo::Redirect& getRedirect() {
+          ASSERT(tag, ==, REDIRECT);
+          return redirect;
+        }
+        DirentInfo::Resolved& getResolved() {
+          ASSERT(tag, ==, RESOLVED);
+          return resolved;
+        }
+        const DirentInfo::Direct& getDirect() const {
+          ASSERT(tag, ==, DIRECT);
+          return direct;
+        }
+        const DirentInfo::Redirect& getRedirect() const {
+          ASSERT(tag, ==, REDIRECT);
+          return redirect;
+        }
+        const DirentInfo::Resolved& getResolved() const {
+          ASSERT(tag, ==, RESOLVED);
+          return resolved;
+        }
+
+      private: // members
+        union {
+          Direct direct;
+          Redirect redirect;
+          Resolved resolved;
+        } PACKED;
+
+      public: // members
+        enum : char {DIRECT, REDIRECT, RESOLVED} tag;
+    } PACKED;
 
     class Dirent
     {
         static const uint16_t redirectMimeType = 0xffff;
         static const uint32_t version = 0;
 
+        PathTitleTinyString pathTitle;
         uint16_t mimeType;
-        DirentInfo info {};
-        char ns;
-        std::string path;
-        std::string title;
-        Cluster* cluster = nullptr;
-        char redirectNs;
-        std::string redirectPath;
         entry_index_t idx = entry_index_t(0);
+        DirentInfo info;
         offset_t offset;
+        char ns;
         bool removed;
```

Comment:
I think it is possible to win one more byte in the size of `Dirent` by getting rid of the `removed` data member (you can mark a removed dirent with a special value in one of its other fields, e.g. `ns='\0'`).

Reply:
That's a interesting idea, and we can even go a bit further by also integrate the front-article information.
But let's delay this for another PR, it is better to keep this one "short".